packages/client/src/compatibility/compatibility.ts

```diff
@@ -0,0 +1,202 @@
+import { RuntimeContext } from "@polkadot-api/observable-client"
+import {
+  CompatibilityCache,
+  CompatibilityLevel,
+  EntryPoint,
+  entryPointsAreCompatible,
+  TypedefNode,
+  valueIsCompatibleWithDest,
+} from "@polkadot-api/metadata-compatibility"
+import { ChainDefinition } from "@/descriptors"
+import {
+  CompatCtx,
+  getDestCompatCtx,
+  getEntryAndGetter,
+  getUserCompatCtx,
+  OpType,
+} from "./compat-ctx"
+import { createProxyPath } from "@/utils"
+import { withWeakCache } from "@/utils/with-weak-cache"
+
+export type CompatApi<T> = Promise<(ctx: RuntimeContext) => T>
+export type CompatHelper<T = any> = {
+  level: CompatibilityLevel
+  isApiCompatible: (from?: CompatibilityLevel) => boolean
+  isValueCompatible: (dest: T) => boolean
+}
+export type ArgsValueCompatHelper<Args = any, Value = any> = {
+  args: CompatHelper<Args>
+  value: CompatHelper<Value>
+  isApiCompatible: (from?: CompatibilityLevel) => boolean
+}
+export type ValueCompat = CompatApi<CompatHelper>
+export type InOutCompat = CompatApi<ArgsValueCompatHelper>
+export type ConstCompat = Promise<(dest: any) => boolean>
+
+const incompatible: CompatHelper = {
+  level: CompatibilityLevel.Incompatible,
+  isApiCompatible: () => false,
+  isValueCompatible: () => false,
+}
+const inOutIncompat: ArgsValueCompatHelper = {
+  args: incompatible,
+  value: incompatible,
+  isApiCompatible: () => false,
+}
+
+const identical: CompatHelper = {
+  level: CompatibilityLevel.Identical,
+  isApiCompatible: () => true,
+  isValueCompatible: () => true,
+}
+
+const getIsApiCompatible =
+  (currentLevel: CompatibilityLevel) =>
+  (target = CompatibilityLevel.BackwardsCompatible) =>
+    currentLevel >= target
+
+const getCompatibilityHelper = <K extends OpType>(
+  kind: OpType,
+  cache: CompatibilityCache,
+  user: {
+    entry?: EntryPoint
+    getter: (id: number) => TypedefNode
+  } | null,
+  dest: {
+    entry?: EntryPoint
+    getter: (id: number) => TypedefNode
+  },
+): K extends OpType.Tx | OpType.Event | OpType.Const
+  ? CompatHelper
+  : ArgsValueCompatHelper => {
+  let result: { args: CompatHelper; value: CompatHelper }
+
+  if (!dest.entry) result = inOutIncompat
+  else {
+    const { args } = dest.entry
+
+    const _isCompat = (value: any) =>
+      valueIsCompatibleWithDest(args, dest.getter, value)
+    const isValueCompatible =
+      kind === OpType.Storage
+        ? (value: any[]) => _isCompat(value.length === 1 ? value[0] : value)
+        : _isCompat
+
+    if (!user?.entry) {
+      const level = CompatibilityLevel.Partial
+      result = {
+        args: {
+          level,
+          isValueCompatible,
+          isApiCompatible: getIsApiCompatible(level),
+        },
+        value: user ? incompatible : identical,
+      }
```

Comment:
I'm still not sure about this. Maybe we have to talk again. As I see it, there are two cases:

1. Unsafe API (no descriptors): In this case, the compatibility API doesn't make too much sense, but I'd say it would have to be consistent - `args` and `value` are both either `Partial` or `Identical`.
2. Typed API, but where the user is doing something weird, accessing an API that was not in the descriptors. Maybe doing `as any` or mixing up stuff. Same thing, I'd say it would have to be consistent, both `args` and `value` either `Partial` or `Incompatible`.